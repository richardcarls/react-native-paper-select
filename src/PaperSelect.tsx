import * as React from 'react';

import { View, Keyboard, Platform } from 'react-native';

import { TextInputAnchor } from './TextInputAnchor';
import { DropdownMenu } from './DropdownMenu';
import { ModalMenu } from './ModalMenu';

import { optionCompare, objectGetProperty } from './util';

export type PaperSelectOption = {
  value: string;
  label: string;
};

export type PaperSelectProps<T extends NonNullable<any>> = {
  // TODO: multi

  // renderInput?: 'text' | 'button' | 'chips' | (() => React.ReactElement);
  renderInput?: 'text';

  // renderMenu?: 'modal' | 'dropdown' | false | (() => React.ReactElement);
  renderMenu?: 'modal' | 'dropdown' | false;

  /** The value to display in the component */
  value?: T;

  defaultValue?: T;

  /** Array of options data */
  options?: ReadonlyArray<T>;

  label?: string;

  // TODO: nullable

  /** Use error styles on the component */
  error?: boolean;

  // TODO: editable

  disabled?: boolean;

  /**
   * Label to use for the optional "none" option (sets value to `undefined`)
   *
   * @defaultValue '(None)'
   */
  noneOption?: String | false;

  /**
   * Callback that is called when the component's selection changes.
   *
   * The selected option is passed as an argument.
   */
  onSelection?: (selected: T | undefined) => void;

  /**
   * Returns a value and label for the given option
   *
   * @param option - The option to derive a value and label from
   *
   * @defaultValue
   * When not defined:
   * - If option is a string, it's value is also used for the label; otherwise
   * - The values of the `value` and `label` properties, if they exist; otherwise
   * - The option is coersed to a string via `String()`
   */
  extractorFn?: (option: T) => PaperSelectOption;

  // TODO: onSelectionCommit

  /** testID to be used on tests. */
  testID?: string;
};

/**
 * A component that allows users to make a selection from a list of options
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { PaperSelect } from '@rcarls/react-native-paper-select';
 *
 * const options = ['one', 'two', 'three'];
 *
 * const MyComponent = () => {
 *   const [value, setValue] = React.useState('');
 *
 *   return (
 *     <PaperSelect
 *       label="Example"
 *       options={options}
 *       value={value}
 *       onSelection={(option) => setValue(option)}
 *     />
 *   );
 * };
 *
 * export default MyComponent;
 * ```
 */
export const PaperSelect = <T extends NonNullable<any>>(
  props: PaperSelectProps<T>
) => {
  const {
    renderInput = 'text',
    // TODO: If undefined, select based on OS env (and if renderInput is not 'chips')
    renderMenu = Platform.OS === 'web' && 'document' in global
      ? 'dropdown'
      : 'modal',
    value: otherValue,
    defaultValue,
    options,
    label,
    error = false,
    disabled = false,
    noneOption = '(None)',
    onSelection,
    extractorFn = defaultExtractorFn,
    testID,
  } = props;

  const [uncontrolledValue, setUncontrolledValue] = React.useState<
    T | undefined
  >(otherValue ?? defaultValue);

  // Use value from props instead of local state when input is controlled
  const isControlled = otherValue !== undefined;
  const value = isControlled ? otherValue : uncontrolledValue;

  const [menuVisible, setMenuVisible] = React.useState(false);

  // TODO: Better keyboard accessibility? (ie: up/down traversal, space open/select, esc close, etc.)

  const onSelectionConfirmed = (selection: T | undefined) => {
    if (
      !isControlled &&
      options?.some((option) => optionCompare(option, value))
    ) {
      // Keep track of value in local state when input is not controlled
      setUncontrolledValue(value);
    }

    onSelection && onSelection(selection);
    setMenuVisible(false);
  };

  const openMenu = React.useCallback(() => {
    if (renderMenu && !disabled) {
      // Close keyboard to make best use of screen space
      Keyboard.dismiss();

      setMenuVisible(true);
    }
  }, [renderMenu, disabled]);

  const Anchor = (
    <>
      {renderInput === 'text' ? (
        <TextInputAnchor
          active={menuVisible}
          label={label}
          value={value ? extractorFn(value).label : ''}
          onPress={() => openMenu()}
          disabled={disabled}
          error={error}
          testID={testID ? `${testID}-anchor` : undefined}
        />
      ) : null}
    </>
  );

  return (
    <View
      accessible={true}
      accessibilityRole="combobox"
      accessibilityLabel={label}
      accessibilityValue={{ text: value ? extractorFn(value).value : '' }}
      accessibilityState={{ disabled, expanded: menuVisible }}
      testID={testID}
    >
      {renderMenu === 'dropdown' ? (
        <DropdownMenu
          options={options}
          selected={value}
          visible={menuVisible}
          noneLabel={noneOption ? noneOption : undefined}
          extractorFn={extractorFn}
          onSelection={onSelectionConfirmed}
          onDismiss={() => setMenuVisible(false)}
          testID={testID ? `${testID}-dropdown` : undefined}
        >
          {Anchor}
        </DropdownMenu>
      ) : null}

      {renderMenu === 'modal' ? (
        <>
          <ModalMenu
            options={options}
            selected={value}
            visible={menuVisible}
            label={label}
            extractorFn={extractorFn}
            onSelection={onSelectionConfirmed}
            onDismiss={() => setMenuVisible(false)}
            testID={testID ? `${testID}-modal` : undefined}
          />
          {Anchor}
        </>
      ) : null}
    </View>
  );
};

export default PaperSelect;

const defaultExtractorFn = <T extends NonNullable<any>>(option?: T) => {
  if (option === undefined) {
    return { value: undefined, label: undefined };
  }

  if (typeof option === 'string') {
    return { value: option, label: option };
  }

  let label, value;
  if (typeof option === 'object' && option !== null) {
    value =
      objectGetProperty(option, 'value') ||
      objectGetProperty(option, 'id') ||
      objectGetProperty(option, 'key');

    label = objectGetProperty(option, 'label');
  }

  value = value ?? String(option);

  return {
    label: label ?? value,
    value,
  };
};
