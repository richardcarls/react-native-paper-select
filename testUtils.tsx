import * as React from 'react';

import { Provider as PaperProvider, MD3LightTheme } from 'react-native-paper';

export const PaperProviderContext = ({ children }: React.PropsWithChildren) => {
  return <PaperProvider theme={MD3LightTheme}>{children}</PaperProvider>;
};
