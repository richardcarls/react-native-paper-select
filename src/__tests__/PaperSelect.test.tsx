import * as React from 'react';
import { render, screen } from '@testing-library/react-native';
import { PaperProviderContext } from '../testUtils';

import { PaperSelect } from '../PaperSelect';

describe('<PaperSelect />', () => {
  it('should match snapshot', () => {
    render(<PaperSelect />, { wrapper: PaperProviderContext });

    expect(screen).toMatchSnapshot();
  });

  it('should exist in the DOM', async () => {
    render(<PaperSelect />, { wrapper: PaperProviderContext });

    const component = await screen.findByRole('combobox');

    expect(component).toBeDefined();
  });
});
