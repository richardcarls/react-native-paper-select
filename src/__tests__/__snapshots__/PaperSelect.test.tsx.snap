// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaperSelect /> should match snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      accessibilityRole="combobox"
      accessibilityState={
        Object {
          "disabled": false,
          "expanded": false,
        }
      }
      accessibilityValue={
        Object {
          "text": "",
        }
      }
      accessible={true}
    >
      <View
        accessibilityState={
          Object {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          Object {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "height": "100%",
            "left": 0,
            "opacity": 0,
            "position": "absolute",
            "top": 0,
            "width": "100%",
          }
        }
      />
      <View
        pointerEvents="none"
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgba(231, 224, 236, 1)",
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
              },
              Object {
                "color": "rgba(28, 27, 31, 1)",
              },
            ]
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "backgroundColor": "rgba(28, 27, 31, 1)",
                "bottom": 0,
                "height": 1,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "transform": Array [
                  Object {
                    "scaleY": 0.5,
                  },
                ],
                "zIndex": 1,
              }
            }
            testID="text-input-underline"
          />
          <View
            style={
              Array [
                Object {
                  "paddingBottom": 0,
                  "paddingTop": 0,
                },
                Object {
                  "minHeight": 56,
                },
              ]
            }
          >
            <TextInput
              editable={false}
              maxFontSizeMultiplier={1.5}
              multiline={false}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholderTextColor="rgba(73, 69, 79, 1)"
              selectionColor="rgba(103, 80, 164, 1)"
              style={
                Array [
                  Object {
                    "margin": 0,
                  },
                  Object {
                    "paddingLeft": 16,
                    "paddingRight": 56,
                  },
                  Object {
                    "height": 56,
                  },
                  Object {
                    "paddingBottom": 0,
                    "paddingTop": 0,
                  },
                  Object {
                    "color": "rgba(73, 69, 79, 1)",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "letterSpacing": 0.15,
                    "lineHeight": undefined,
                    "textAlign": "left",
                    "textAlignVertical": "center",
                  },
                  false,
                  Object {
                    "marginRight": 40,
                    "paddingRight": 16,
                  },
                  undefined,
                ]
              }
              testID="text-input-flat"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 24,
                  "justifyContent": "center",
                  "position": "absolute",
                  "width": 24,
                },
                Object {
                  "right": 16,
                  "top": 16,
                },
              ]
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "alignSelf": undefined,
                  "backgroundColor": "transparent",
                  "borderRadius": 20,
                  "bottom": undefined,
                  "end": undefined,
                  "flex": undefined,
                  "height": 40,
                  "left": undefined,
                  "margin": 0,
                  "opacity": undefined,
                  "position": undefined,
                  "right": undefined,
                  "shadowColor": "#000",
                  "shadowOffset": Object {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0,
                  "shadowRadius": 0,
                  "start": undefined,
                  "top": undefined,
                  "transform": undefined,
                  "width": 40,
                }
              }
              testID="right-icon-adornment-container-outer-layer"
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "borderColor": "rgba(28, 27, 31, 0.12)",
                    "borderRadius": 20,
                    "borderWidth": 0,
                    "elevation": 0,
                    "flex": 1,
                    "overflow": "hidden",
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 0,
                      "width": 0,
                    },
                    "shadowOpacity": 0,
                    "shadowRadius": 0,
                  }
                }
                testID="right-icon-adornment-container"
              >
                <View
                  accessibilityComponentType="button"
                  accessibilityRole="none"
                  accessibilityState={
                    Object {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": true,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityTraits={
                    Array [
                      "button",
                      "disabled",
                    ]
                  }
                  accessibilityValue={
                    Object {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  centered={true}
                  collapsable={false}
                  focusable={true}
                  hitSlop={
                    Object {
                      "bottom": 6,
                      "left": 6,
                      "right": 6,
                      "top": 6,
                    }
                  }
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Array [
                      Object {
                        "overflow": "hidden",
                      },
                      false,
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexGrow": 1,
                          "justifyContent": "center",
                        },
                        Object {
                          "borderRadius": 20,
                        },
                      ],
                    ]
                  }
                  testID="right-icon-adornment"
                >
                  <Text
                    accessibilityElementsHidden={true}
                    allowFontScaling={false}
                    importantForAccessibility="no-hide-descendants"
                    pointerEvents="none"
                    selectable={false}
                    style={
                      Array [
                        Object {
                          "color": "rgba(28, 27, 31, 0.38)",
                          "fontSize": 24,
                        },
                        Array [
                          Object {
                            "lineHeight": 24,
                            "transform": Array [
                              Object {
                                "scaleX": 1,
                              },
                            ],
                          },
                          Object {
                            "backgroundColor": "transparent",
                          },
                        ],
                        Object {
                          "fontFamily": "Material Design Icons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        Object {},
                      ]
                    }
                  >
                    󰅀
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</RNCSafeAreaProvider>
`;
